pkg/downsample/downsample.go

```diff
@@ -0,0 +1,178 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package downsample
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/timescale/promscale/pkg/internal/day"
+)
+
+const (
+	setDownsamplingStateSQL       = "SELECT prom_api.set_downsampling_state($1)"
+	createDownsamplingSQL         = "CALL _prom_catalog.create_downsampling($1, $2, $3)"
+	updateDownsamplingStateForSQL = "SELECT _prom_catalog.update_downsampling_state($1, $2)"
+	downsamplePrefix              = "ds_" // Stands of downsample_
+)
+
+type Config struct {
+	Interval  day.Duration `yaml:"interval"`
+	Retention day.Duration `yaml:"retention"`
+}
+
+func (c Config) Name() string {
+	return downsamplePrefix + c.Interval.Text()
+}
+
+func SetState(ctx context.Context, conn *pgx.Conn, state bool) error {
+	_, err := conn.Exec(ctx, setDownsamplingStateSQL, state)
+	if err != nil {
+		return fmt.Errorf("error setting downsampling state: %w", err)
+	}
+	return nil
+}
+
+type definition struct {
+	interval, retention time.Duration
+	disabled            bool
+}
+
+// Sync updates the downsampling cfgs in the DB in accordance with the given new cfgs. It:
+//  1. Creates of new downsampling cfgs that are not in the database
+//  2. Updates retention duration of downsampling cfgs that are present in the database but with a different retention duration
+//  3. Disables refreshing of downsampling cfgs in the database that were not found in the new cfgs
+//  4. Enables refreshing of downsampling cfgs in the database that are in the new cfgs but were previously disabled
+func Sync(ctx context.Context, conn *pgx.Conn, cfgs []Config) error {
```

Comment:
I feel we can simplify this logic by merging two sets in memory: one is array of new `Config`, another is array of existing `Config`. Upon merge we just UPSERT Configs (insert or update existing). Having a `merge` method defined will also allow for easier unit testing.

Reply:
is this correct?

1. if existing config is missing in new set, disable it
2. if new config is not existing, insert
3. if new config matches existing config on schema name, update it UNLESS the interval is different, in which case, throw an error
4. no two configs can have the same interval
5. no two configs can have the same name

Reply:
honestly, I'd write a single database function to handle this. It's much fewer lines of code and one database round-trip.

something like this:
```sql
create or replace function apply_downsample_config(config jsonb) returns void
as $proc$
declare
    _input _prom_catalog.downsample[];
    _schema_name text;
begin
    select array_agg(x) into strict _input
    from jsonb_populate_recordset(null::_prom_catalog.downsample, config) x;

    for _schema_name in
    (
        select x.schema_name
        from unnest(_input) x
        where not exists
        (
            select 1
            from _prom_catalog.downsample d
            where d.schema_name = x.schema_name
        )
    )
    loop
        execute format('create schema %I', _schema_name);
    end loop;

    -- insert or update
    insert into _prom_catalog.downsample (schema_name, resolution, retention, should_refresh)
    select x.schema_name, x.resolution, x.retention, true
    from unnest(_input) x
    on conflict (schema_name) do
    update set
        retention = excluded.retention,
        should_refresh = true
    ;

    -- disable if missing
    update _prom_catalog.downsample set should_refresh = false
    where schema_name not in 
    (
        select schema_name
        from unnest(_input)
    );
end;
$proc$ language plpgsql;
```

then you can just push the new config to the database as json. almost no logic required on the go side

```sql
select apply_downsample_config(
$$
[
    {
        "schema_name": "d_1d",
        "resolution": "1d",
        "retention": "30d"
    },
    {
        "schema_name": "d_12h",
        "resolution": "12h",
        "retention": "18d"
    },
    {
        "schema_name": "d_1h",
        "resolution": "1h",
        "retention": "15d"
    }
]
$$::jsonb
);
```

Reply:
I am not confident for pushing down the entire config for 3 reasons:
1. The point of connector applying the configuration is that in future, if the `downsample.Config` needs more fields, we can easily change there and connector handles it. With `apply_downsample_config`, I don't see how we can do unless we make changes in the extension code.
2. This can be specific to me. I wrote SQL code so it's easy to read and understand (I am not an expert in SQL). In the current SQL code, I tried to keep everything very straightforward since for me, I find following SQL not as easy as Golang. I am thankful for the new suggestion, but I think `apply_downsample_config` is not a straightforward function. Plus, if there happens to be any changes in `downsample.Config` in future, we cannot simply push the jsonb until we filter some fields, since the `_prom_catalog.downsample` will not have the additional fields in future. I feel this is creating a strong coupling.
3. With `apply_downsample_config()`, the SQL API is no longer simple. It does more than 1 task and I am not sure if its fine.

These are my thoughts. I do not want to oppose it, as I feel I am not the right person in SQL to suggest. So if everyone is happy with `apply_downsample_config` then I can write this function.

Reply:
My 2 cents. I like the approach of the DB function:

1- If we add more fields to the config we would still need to store them in the DB, right?. This is just to handle the logic of applying the whole config. I find this simpler than doing the whole code in the connector and managing the logic there. For me, it feels less error prone and doesn't seem like an out of place extension to the downsample API.

2- 
> I am not an expert in SQL

This is precisely why you should try to write it. You'll have the guidance of the awesome data-platform team. They'll review and help you, and get you a little bit closer to being a SQL expert.

If you don't feel confortable working with the JSONB because of the loose schema, maybe there are alternatives. Could we have the config be a custom type and just pass an array of those, or something, just an idea I'm not a PostgreSQL expert either 😅 

3- You could say that this function does 1 thing, it replaces the current config with the given config.

Reply:
Sorry, my point was more towards keeping things simple and straightforward. That will make life easier for all of us. Plus, keeping things in Golang allows us to test easily.

Reply:
There are more points than just these. If we make create/update/disable in extension, then I will have to remove `TestMetricDownsampleSync` completely (and write in SQL). Otherwise, it will be creation/updation in extension, and its tests in Go. Then, I will have to write tests in SQL and it will be a very big PR to review. Given we are not sure of this feature's impact ATM, I don't feel confident investing more time and designing more. We should probably go now with speed than accuracy.